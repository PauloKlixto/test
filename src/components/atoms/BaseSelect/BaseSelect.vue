<template>
  <div class="BaseSelect" :class="componentClasses">
    <input
      ref="BaseSelect__input"
      v-bind="$attrs"
      type="hidden"
      name="BaseSelect__input"
    />

    <label class="BaseSelect__label" :class="BaseSelectLabelClasses">
      {{ label }}
    </label>
    <div class="BaseSelect__container" :class="{ open: menuOpen }">
      <div
        class="BaseSelect__selectedContainer"
        @click.prevent.stop="toggleSelect"
      >
        <div class="BaseSelect__selected" :title="value.name">
          {{ value.name }}
        </div>
      </div>

      <select v-if="isMobile" class="BaseSelect__select-mobile">
        <option
          v-for="option in options"
          :id="option.id"
          :key="option.id"
          :value="option"
          @click="setValue(option)"
        >
          {{ option.name }}
        </option>
      </select>
      <div v-else ref="items" class="BaseSelect__Items">
        <div
          v-for="option in options"
          :key="option.id"
          class="BaseSelect__Item"
          :title="option.name"
          @click="setValue(option)"
        >
          {{ option.name }}
        </div>
      </div>
    </div>
    <p v-if="isInvalid" class="BaseSelect__invalidLabel">
      {{ invalidLabel }}
    </p>
  </div>
</template>

<script>
import { mapGetters } from 'vuex'

export default {
  name: 'BaseSelect',
  props: {
    label: {
      type: String,
      default: ''
    },
    options: {
      type: Array,
      required: true,
      validator: items =>
        items.filter(item => {
          return (
            Object.prototype.hasOwnProperty.call(item, 'id') &&
            Object.prototype.hasOwnProperty.call(item, 'name')
          )
        }).length === items.length
    },
    value: {
      type: [Object, String, Number],
      required: true
    },
    isInvalid: {
      type: Boolean,
      default: false
    },
    invalidLabel: {
      type: String,
      default: ''
    }
  },
  data() {
    return {
      menuOpen: false,
      maxLength: 17,
      title: ''
    }
  },
  computed: {
    ...mapGetters({
      isMobile: 'userAgent/isMobile'
    }),

    componentClasses() {
      return { 'BaseSelect--invalid': this.isInvalid }
    },

    BaseSelectLabelClasses() {
      return { 'BaseSelect__label--upwards': !!this.value || this.menuOpen }
    }
  },
  watch: {
    // Se o valor recebido n√£o for um objeto, transforma em objeto
    value(val) {
      if (typeof val === 'string' || typeof val === 'number')
        this.$emit('input', { id: val, name: val })
    }
  },

  methods: {
    toggleSelect() {
      if (this.options.length === 1) return
      this.menuOpen ? this.close() : this.open()
    },

    open() {
      window.addEventListener('click', this.toggleSelect)
      this.menuOpen = true
      let h = 108
      this.$refs.items.style.maxHeight = h + 'px'
    },

    close() {
      window.removeEventListener('click', this.toggleSelect)
      this.menuOpen = false
      this.$refs.items.style.maxHeight = '0px'
    },

    setValue(option) {
      if (!this.menuOpen) return

      this.$emit('input', option)
      this.$emit('blur', option)
      this.close()
    }
  }
}
</script>

<style lang="scss" scoped>
.BaseSelect {
  &[disabled] {
    opacity: 0.5;
    cursor: not-allowed;
    & > * {
      pointer-events: none;
    }
  }
  &__label {
    font-size: 17px;
    color: $color-light;
    transform: translateY(10px);
    display: block;

    &--upwards {
      font-size: 13px;
      transform: translateY(0px);
    }
  }
  &__container {
    position: relative;
    border-bottom: 1px solid $color-light;
    box-sizing: border-box;

    &:before {
      @include chevronAlt($color-red);
      right: 5px;
      bottom: 10px;
      transform: rotateZ(225deg);
    }

    &.open {
      &:before {
        transform: rotateZ(45deg);
      }
    }

    .BaseSelect--invalid & {
      border-bottom: 1px solid $color-red;
      color: $color-red;
    }
  }
  &__invalidLabel {
    padding: 5px 10px;
    font-size: 12px;
    color: $color-red;
    font-weight: 900;
    position: absolute;
  }
  &__selectedContainer {
    display: flex;
    align-items: center;
    justify-content: space-between;
    padding-top: 10px;
    padding-bottom: 5px;
    padding-left: 10px;
    cursor: pointer;
  }
  &__selected {
    font-weight: 100;
    font-size: 0.9rem;
    white-space: nowrap;
    max-width: 200px;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
  }
  &__Items {
    position: absolute;
    width: calc(100% + 2px);
    left: 0;
    top: 100%;
    border: 1px solid $color-light;
    border-top: none;
    height: auto;
    max-height: 0;
    overflow-y: auto;
    transition: max-height 0.2s;
    transition-timing-function: linear;
    z-index: $z-dropdown;

    .BaseSelect--invalid & {
      border: none;
    }
  }
  &__Item {
    padding: 5px 15px;
    font-weight: 100;
    font-size: 0.9rem;
    cursor: pointer;
    background-color: #fff;
    text-transform: initial;
    white-space: nowrap;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    &:hover {
      background-color: $color-light;
    }
    &:first-child {
      margin-top: 2px;
    }
  }
  &__select-mobile {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    opacity: 0;
  }
}
</style>
