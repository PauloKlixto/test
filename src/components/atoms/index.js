import BaseBackButton from './BaseBackButton/BaseBackButton'
import BaseButton from './BaseButton/BaseButton'
import BaseCreditCard from './BaseCreditCard/BaseCreditCard'
import BaseHeading from './BaseHeading/BaseHeading'
import BaseSelect from './BaseSelect/BaseSelect'
import BaseSteps from './BaseSteps/BaseSteps'

export {
  BaseBackButton,
  BaseButton,
  BaseCreditCard,
  BaseHeading,
  BaseSelect,
  BaseSteps
}
