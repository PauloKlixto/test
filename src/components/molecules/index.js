import InputGroup from './InputGroup/InputGroup'
import DynamicForm from './DynamicForm/DynamicForm'

export { InputGroup, DynamicForm }
