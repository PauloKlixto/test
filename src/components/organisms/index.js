import TheHeader from './TheHeader/TheHeader'

export { TheHeader }
