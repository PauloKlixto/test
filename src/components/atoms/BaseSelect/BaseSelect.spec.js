/* eslint-disable no-undef */
import { shallowMount, createLocalVue } from '@vue/test-utils'
import { Icon } from '@/components/photons'
import BaseSelect from './BaseSelect'
import optionsData from './utils/options'
import Vuex from 'vuex'

const localVue = createLocalVue()
localVue.use(Vuex)

localVue.component('Icon', Icon)

describe('BaseSelect', () => {
  let store

  store = new Vuex.Store({
    modules: {
      userAgent: {
        namespaced: true,
        getters: {
          isMobile: () => true
        }
      }
    }
  })

  const WRAPPER_MOBILE = shallowMount(BaseSelect, {
    localVue,
    store,
    propsData: { options: optionsData, value: {} }
  })

  store = new Vuex.Store({
    modules: {
      userAgent: {
        namespaced: true,
        getters: {
          isMobile: () => false
        }
      }
    }
  })

  const WRAPPER = shallowMount(BaseSelect, {
    localVue,
    store,
    propsData: { options: optionsData, value: {} }
  })

  test('Componente existe', () => {
    expect(WRAPPER.exists()).toBe(true)
  })

  test('Componente abre', () => {
    WRAPPER.vm.open()

    const IS_OPEN = WRAPPER.vm.menuOpen
    expect(IS_OPEN).toBeTruthy()
  })

  test('Componente fecha', () => {
    WRAPPER.vm.close()
    const IS_OPEN = WRAPPER.vm.menuOpen
    expect(IS_OPEN).toBeFalsy()
  })

  test('Click item de select', () => {
    let w = WRAPPER.find('.BaseSelect__Item')
    w.trigger('click')

    const IS_OPEN = WRAPPER.vm.menuOpen
    expect(IS_OPEN).toBeFalsy()
  })

  test('Click no item selecionado [true]', () => {
    WRAPPER.vm.menuOpen = true
    const IS_OPEN = WRAPPER.vm.menuOpen

    WRAPPER.find('.BaseSelect__selectedContainer').trigger('click')
    expect(IS_OPEN).toBeTruthy()
  })

  test('Click no item selecionado [false]', () => {
    WRAPPER.vm.menuOpen = false
    const IS_OPEN = WRAPPER.vm.menuOpen

    WRAPPER.find('.BaseSelect__selectedContainer').trigger('click')
    expect(IS_OPEN).toBeFalsy()
  })

  test('Select Item vers達o mobile', () => {
    WRAPPER_MOBILE.find('.BaseSelect__select-mobile').trigger('change')
  })

  describe('Computed - BaseSelectLabelClasses', () => {
    test('Quando tem value ou menuOpen true', () => {
      WRAPPER.setData({ menuOpen: true })
      WRAPPER.setProps({ value: { id: 0, name: 'Teste' } })
      expect(WRAPPER.vm.BaseSelectLabelClasses).toStrictEqual({
        'BaseSelect__label--upwards': true
      })
    })

    test('Quando tem value ou menuOpen false', () => {
      WRAPPER.setData({ menuOpen: false })
      WRAPPER.setProps({ value: { id: 0, name: 'Teste' } })
      expect(WRAPPER.vm.BaseSelectLabelClasses).toStrictEqual({
        'BaseSelect__label--upwards': true
      })
    })

    test('Quando n達o tem value ou menuOpen true', () => {
      WRAPPER.setData({ menuOpen: true })
      WRAPPER.setProps({ value: '' })
      expect(WRAPPER.vm.BaseSelectLabelClasses).toStrictEqual({
        'BaseSelect__label--upwards': true
      })
    })

    test('Quando n達o tem value ou menuOpen false', () => {
      WRAPPER.setData({ menuOpen: false })
      WRAPPER.setProps({ value: '' })
      expect(WRAPPER.vm.BaseSelectLabelClasses).toStrictEqual({
        'BaseSelect__label--upwards': false
      })
    })
  })

  test('watch - value', () => {
    WRAPPER.setProps({ value: false })
    WRAPPER.setProps({ value: true })
  })

  test('methods - toggleSelect', () => {
    WRAPPER.setProps({ options: [{}] })
    WRAPPER.vm.toggleSelect()

    WRAPPER.setProps({ options: [] })
    WRAPPER.vm.toggleSelect()
  })

  test('methods - setValue', () => {
    WRAPPER.vm.menuOpen = true
    WRAPPER.vm.setValue('option')
    expect(WRAPPER.emitted('input')).toBeTruthy()
    expect(WRAPPER.emitted('blur')).toBeTruthy()

    WRAPPER.vm.menuOpen = false
    WRAPPER.vm.setValue('option')
  })

  test('Componente n達o alterado', () => {
    expect(WRAPPER.html()).toMatchSnapshot()
  })
})
