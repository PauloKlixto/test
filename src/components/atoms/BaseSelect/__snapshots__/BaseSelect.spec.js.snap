// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseSelect Componente n√£o alterado 1`] = `
"<div class=\\"BaseSelect\\"><input type=\\"hidden\\" name=\\"BaseSelect__input\\"> <label class=\\"BaseSelect__label BaseSelect__label--upwards\\">

  </label>
  <div class=\\"BaseSelect__container\\">
    <div class=\\"BaseSelect__selectedContainer\\">
      <div class=\\"BaseSelect__selected\\">

      </div>
    </div>
    <div class=\\"BaseSelect__Items\\" style=\\"max-height: 0px;\\"></div>
  </div>
  <!---->
</div>"
`;
